Add the saltext history, describe the module extraction process (#1)
lkubb/saltext-copier: docs/topics/extraction.md

# Extraction of Salt core modules

Below are some rough steps that extract an existing set of modules into an extension while preserving the Git history. Let's use the `stalekey` engine as an example.

## A module extraction example

### 1. Install the Git history filtering tool

```shell
pip install git-filter-repo
```

### 2. Clone the Salt repo and analyze its history

```shell
git clone https://github.com/saltstack/salt --single-branch
cd salt
git filter-repo --analyze
tree .git/filter-repo/analysis/

> rg stalekey .git/filter-repo/analysis/path-all-sizes.txt

Reviewer: question: Is ripgrep required or does grep work just as fine? If the former, I think we should a Prerequisites heading and tell people to install it.

Author: Grep will work just fine, ripgrep is just what I had in my notes.

Reviewer: Let's `%s/rg/grep/g` then for minimalism's sake :)

### 3. Filter the history into a separate branch

```shell
git checkout -b filter-source
git filter-repo \
    --path salt/engines/stalekey.py \
    --path-rename salt/engines/stalekey.py:src/saltext/stalekey/engines/stalekey.py \
    --path tests/pytests/unit/engines/test_stalekey.py \
    --path tests/unit/engines/test_stalekey.py \
    --refs refs/heads/filter-source --force
```

### 4. Clean up the history

```shell
git log --name-only
git rebase -i --root
```

The main goal here is to remove the commits that do not touch any of the extracted files, and also delete the last commit that removes them. The merge commits seem to be removed automatically during the rebase.

While looking at the Git log, please note the major contributors (to add them as code authors later).

### 5. Populate the extension repo

When answering the Copier questions, choose the `engine` module type only, specify yourself as an author:

```shell
cd ..
mkdir saltext-stalekey && cd saltext-stalekey
git init
copier copy --trust https://github.com/salt-extensions/salt-extension-copier ./
```

We need to remove some boilerplate files generated by Copier:

```shell

> rm -rf tests/functional tests/integration tests/unit/engines/test_stalekey.py src/saltext/stalekey/engines/stalekey_mod.py

Reviewer: ```suggestion
rm -f tests/**/test_*.py src/**/*_mod.py
```

suggestion: I would prefer if people would just delete the boilerplate intended to help write a new extension. `tests/functional/conftest.py` and `tests/integration/conftest.py` should be kept in case someone wants to add such tests in the future. They will also receive updates through Copier, unlike the above paths.

```

And then merge the history:

```shell
git remote add repo-source ../salt
git fetch repo-source
git merge repo-source/filter-source
git remote rm repo-source
git tag | xargs git tag -d
```

### 6. Clean up and testing

Move the tests into the right location:

```shell

> git mv tests/pytests/unit/engines/test_stalekey.py tests/unit/engines

Reviewer: questions:

* Earlier you mention that `tests/unit/engines/test_stalekey.py` also exists. Doesn't this delete the non-Pytest tests?
* Do you know if non-Pytest tests work with the default saltext test suite? Never tried it, but assumed it would fail unless the whole translation layer was migrated from Salt core. Maybe we should add a note about conversion of tests and how to skip the whole test module to be converted at a later date.

Author: The non-pytest test no longer exists, it was removed even before the great purge PR. I included it in the filtered history for the sake of completeness (maybe someone will want to run git blame that far, to pre-pytest times).

> Do you know if non-Pytest tests work with the default saltext test suite?

I do not know this.

rm -rf tests/pytests
```

Run the automatic fixups:

```shell
pip install git+https://github.com/saltstack/salt-rewrite
SALTEXT_NAME=stalekey salt-rewrite -F fix_saltext .
```

```shell
pip install -e ".[dev,tests,docs]"

> pre-commit run -a  # make sure it is happy

Reviewer: ```suggestion
# Ensure you're inside your project's virtual environment from here on.
python -m venv venv --prompt saltext-stalekey
source venv/bin/activate
python -m pip install -e ".[dev,tests,docs]"
python -m pre_commit run -a   # make sure it is happy
python -m pre_commit install  # ensure the pre-commit hook runs when committing in the future
```

suggestion: I think we should point out to run this inside a venv.

git status
git add .
git commit -m 'Add extension layout'
```

Add the main authors to pyproject.toml:

```shell
vi pyproject.toml
git add pyproject.toml
git commit -m 'Add authors'
```

> Try running the test suite locally until it passes, then commit and push it to run the full test suite on GitHub.

Reviewer: ```suggestion
Try [running the test suite](run-tests-target) and [building the docs](build-docs-target) locally until both pass, then commit and push it to run the full test suite on GitHub.
```

suggestion: The docs are often an issue as well since the default configuration introduces a link checker.
